flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/HITSAlgorithm.java

```diff
@@ -122,22 +179,45 @@ public void updateVertex(Vertex<K, Tuple2<DoubleValue, DoubleValue>> vertex, Mes
 			DoubleValue newAuthorityValue = vertex.getValue().f1;
 
 			if (getSuperstepNumber() > 1) {
-				iterationValueSum = Math.sqrt(((DoubleValue) getPreviousIterationAggregate("sumVertexValue")).getValue());
+				iterationValueSum = Math.sqrt(((DoubleValue) getPreviousIterationAggregate("updatedValueSum")).getValue());
 			}
-			if (getSuperstepNumber() != maxIteration) {
+			if (getSuperstepNumber() < maxIteration) {
 				if (getSuperstepNumber() % 2 == 1) {
-					newHubValue.setValue(newHubValue.getValue() / iterationValueSum);
-					newAuthorityValue.setValue(updateValue);
+
+					//in the first iteration, the diff is the authority value of each vertex
+					double previousAuthAverage = 1.0;
+					double diffValueSum = 1.0 * numberOfVertices;
+					if (getSuperstepNumber() > 1) {
+						previousAuthAverage = ((DoubleValue) getPreviousIterationAggregate("authorityValueSum")).getValue() / numberOfVertices;
+						diffValueSum = ((DoubleValue) getPreviousIterationAggregate("diffValueSum")).getValue();
+					}
+					authoritySumAggregator.aggregate(previousAuthAverage);
+					
+					if (diffValueSum > convergeThreshold) {
+						newHubValue.setValue(newHubValue.getValue() / iterationValueSum);
+						newAuthorityValue.setValue(updateValue);
+					} else {
+
+						//scores are converged and stop iteration
+						maxIteration = getSuperstepNumber();
```

Comment:
This line can stop the iteration after last vertex updating(final updating of hub normalization). If drop this line, the iteration will go on until `getSuperstep == maxIteration` because there are always some vertices can be updated.